[wip] Adds disputer information and guide. (#699)
filecoin-project/filecoin-docs: docs/mine/lotus/disputer.md

---
title: Disputer

> description: Verifying submitted Window PoSts is expensive, and that cost can drastically increase when network congestion causes the gas base fee to rise. To address this, Filecoin Improvement Proposal allows miners to optimistically accept Window PoSts on-chain without verification, allowing them to be disputed later by off-chain-verifiers.

Reviewer: - Specifically, FIP-0010
- It doesn't so much "allow" for optimistic acceptance, it makes it the behaviour for everyone
- The change isn't for miners to start optimistically accepting PoSts, it's for all node operators

Author: ```suggestion
description: Verifying submitted Window PoSts is expensive, and that cost can drastically increase when network congestion causes the gas base fee to rise. To address this, Filecoin Improvement Proposal FIP-0010 enables node operators to optimistically accept Window PoSts on-chain without verification, allowing them to be disputed later by off-chain-verifiers.
```

The Filecoin Improvement Proposal 0010 (FIP0010) allows miners to _optimistically_ accept Window Proof of Spacetime proofs (Windows PoSt) on-chain without verification, allowing them to be disputed later by off-chain verifiers. Any Lotus node may dispute any on-chain storage proofs submitted in the past 1800 epochs (~15h) by invoking `DisputeWindowedPoSt`.

When a dispute successfully refutes an optimistically accepted Window PoSt, the miner is fined one invalid proof fee (IPF) per active sector in the partition at the moment when said miner submitted the proof, plus a flat fee of 20 FIL. All incorrectly proved sectors are marked faulty, and the address that submitted the dispute is awarded a fixed `DipsuteReward`.

## Penalties and rewards

The penalty for submitting an invalid Window PoSt and the reward for submitting a valid dispute are both subject to change. At the time of writing, those values are:

| Fee/Reward              | Value                                                      |
| ----------------------- | ---------------------------------------------------------- |

> | Invalid proof fee (IPF) | `5.51 *` the expected block reward per sector in 24 hours. |

Author: ```suggestion
| Invalid proof fee (IPF) | (5.51 * the expected block reward per sector in 24 hours) + 20 FIL |
```

> | Valid dispute reward    | 4 FIL                                                      |

Author: @jennijuju are these correct?

> | Valid dispute reward    | 4 FIL                                                      |

Author: @jennijuju is this section still correct?

Reviewer: Yes, see https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0010.md

## Run the disputer

> You can run the disputer by using the `chain disputer` command within Lotus.

Author: ```suggestion
You can run the disputer by using the `chain disputer` command within Lotus. The following examples were ran on a local devnet.
```

> You can run the disputer by using the `chain disputer` command within Lotus. The following examples were ran on a local devnet.

Reviewer: ```suggestion
You can run the disputer by using the `chain disputer` command within Lotus. The following examples were run on a local devnet.
```

1. Fully sync a Lotus node.

> 1. Call the `chain disputer` function with the following variables, following by the `stat` command:

Reviewer: - is "variables" the right term? input params? idk.
- typo: `start`

Author: ```suggestion
1. Call the `chain disputer` function with the following options, following by the `start` command:
```

   | ----------- | ------------------------------------------------------------------------------------------------------------------- |
   | `--max-fee` | The maximum amount of `FIL` that you are willing to spend for a `DisputeWindowedPoSt` message.                      |
   | `--from`    | The address you want to send the messages from. If an address is not specified, Lotus will use the default address. |

   ```shell
   lotus chain disputer --max-fee 1 --from t3r25povzrwpomqlwtxtt25ou76galexvgr3ucgvvtwxiwy2gtqltlzshmtdyz4ys7mt5phoouedengajltbka start

   > checking sync status
   > Worker: 101; Base: 0; Target: 0 (diff: 0)
   > State: <unknown: 0>; Current Epoch: 0; Todo: 0
   > Validated 4 messages (1 per second)
   ```

1. The disputer will continue to verify submitted WinPoSts. To end the disputer press `CTRL` + `c`:

   ```shell
   > Exit by user
   > setting up window post disputer
   > 2021-02-22T17:44:11.362-0500	ERROR	rpc	go-jsonrpc@v0.1.2/client.go:392	got rpc message with cancelled context: context canceled
   > ERROR: Notify stream was invalid
   ```

   You can safely ignore any stream errors when exiting the disputer.

> ## Manual dispute

Reviewer: I wouldn't include info about disputing specific PoSts tbh -- this is an advanced command that's only useful if you a) know what you're doing in detail, and b) have already concluded that a specific PoSt is bad...somehow.

We should probably add some sort of `(advanced users)` note to this command in Lotus.

Author: ```suggestion
## Advanced manual dispute
```

Just because it's advanced doesn't mean it shouldn't be documented. Even if only 1 in 1000 users finds this useful, it's worth having.

> You can send a specific dispute message by running `.lotus chain disputer dispute [minerAddress] [index] [postIndex]`, where:

Author: ```suggestion
You can manually send a specific dispute message using the `dispute` command. This feature has a very narrow use-case and is for advanced users only. To perform a manual dispute run `lotus chain disputer dispute [minerAddress] [index] [postIndex]`, where:
```
